components/scream/src/physics/shoc/shoc_update_prognostics_implicit_impl.hpp

```diff
@@ -13,14 +13,159 @@ namespace shoc {
+void Functions<S,D>::update_prognostics_implicit(
+  const MemberType&            team,
+  const Int&                   nlev,
+  const Int&                   nlevi,
+  const Int&                   num_tracer,
+  const Scalar&                dtime,
+  const uview_1d<const Spack>& dz_zt,
+  const uview_1d<const Spack>& dz_zi,
+  const uview_1d<const Spack>& rho_zt,
+  const uview_1d<const Spack>& zt_grid,
+  const uview_1d<const Spack>& zi_grid,
+  const uview_1d<const Spack>& tk,
+  const uview_1d<const Spack>& tkh,
+  const Scalar&                uw_sfc,
+  const Scalar&                vw_sfc,
+  const Scalar&                wthl_sfc,
+  const Scalar&                wqw_sfc,
+  const uview_1d<const Spack>& wtracer_sfc,
+  const uview_1d<Spack>&       rdp_zt,
+  const uview_1d<Spack>&       tmpi,
+  const uview_1d<Spack>&       tkh_zi,
+  const uview_1d<Spack>&       tk_zi,
+  const uview_1d<Spack>&       rho_zi,
+  const uview_1d<Scalar>&      du,
+  const uview_1d<Scalar>&      dl,
+  const uview_1d<Scalar>&      d,
+  const uview_2d<Spack>&       X1,
+  const uview_2d<Spack>&       X2,
+  const uview_1d<Spack>&       thetal,
+  const uview_1d<Spack>&       qw,
+  const uview_2d<Spack>&       tracer,
+  const uview_1d<Spack>&       tke,
+  const uview_1d<Spack>&       u_wind,
+  const uview_1d<Spack>&       v_wind)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  const auto last_nlev_pack = (nlev-1)/Spack::n;
+  const auto last_nlev_indx = (nlev-1)%Spack::n;
+  const auto last_nlevi_pack = (nlevi-1)/Spack::n;
+  const auto last_nlevi_indx = (nlevi-1)%Spack::n;
+
+  // linearly interpolate tkh, tk, and air density onto the interface grids
+  linear_interp(team,zt_grid,zi_grid,tkh,tkh_zi,nlev,nlevi,0);
+  linear_interp(team,zt_grid,zi_grid,tk,tk_zi,nlev,nlevi,0);
+  linear_interp(team,zt_grid,zi_grid,rho_zt,rho_zi,nlev,nlevi,0);
+
+  // Define the tmpi variable, which is really dt*(g*rho)**2/dp
+  // at interfaces. Substitue dp = g*rho*dz in the above equation
+  compute_tmpi(team, nlevi, dtime, rho_zi, dz_zi, tmpi);
+
+  // compute 1/dp term, needed in diffusion solver
+  dp_inverse(team, nlev, rho_zt, dz_zt, rdp_zt);
+  team.team_barrier();
+
+  // compute terms needed for the implicit surface stress (ksrf)
+  // and tke flux calc (wtke_sfc)
+  Scalar ksrf, wtke_sfc;
+  {
+    const auto wsmin = 1;
+    const auto ksrfmin = 1e-4;
```

Comment:
If you used `const Real XXX = literal;` instead of `const auto` some of the `sp()` calls below could be removed.